src/modules/silentpayments/main_impl.h

```diff
@@ -0,0 +1,713 @@
+
+    return 1;
+}
+
+/** A explanation of the public_data object and its usage:
+ *
+ *  The public_data object contains:
+ *
+ *  [magic: 4 bytes][boolean: 1 byte][A_sum: 64 bytes][input_hash: 32 bytes]
+ *
+ *  The magic bytes are checked by functions using the public_data object to
+ *  check that the public data object was initialized correctly.
+ *
+ *  The boolean (combined) indicates whether or not the A_sum point and the
+ *  input_hash scalar have already been combined or are both included. The reason
+ *  for keeping input_hash and A_sum separate is so that an elliptic curve multiplication
+ *  can be avoided when creating the shared secret, i.e., (b_scan * input_hash) * A_sum.
+ *
+ *  But when storing the public data object, either to send to light clients or for
+ *  wallet rescans, we can save 32-bytes by combining the input_hash and A_sum and saving
+ *  the resulting point serialized as a compressed public key, i.e., input_hash * A_sum.
+ *
+ *  For the each function:
+ *
+ *  - `_recipient_public_data_create` always creates a public_data object with combined = false
+ *  - `_recipient_public_data_serialize` only accepts a public_data object with combined = false
+ *    and then performs an EC mult before serializing the resulting public key as a compressed
+ *    public key
+ *  - `_recpient_public_data_parse` assumes the input represents a previously serialized
+ *    public_data object and always deserializes into a public_data object with combined = true
+ *    (and the input_hash portion zeroed out).
+ */
+
+int secp256k1_silentpayments_recipient_public_data_create(
+    const secp256k1_context *ctx,
+    secp256k1_silentpayments_recipient_public_data *public_data,
+    const unsigned char *outpoint_smallest36,
+    const secp256k1_xonly_pubkey * const *xonly_pubkeys,
+    size_t n_xonly_pubkeys,
+    const secp256k1_pubkey * const *plain_pubkeys,
+    size_t n_plain_pubkeys
+) {
+    size_t i;
+    size_t pubkeylen = 64;
+    secp256k1_ge A_sum_ge, addend;
+    secp256k1_gej A_sum_gej;
+    secp256k1_scalar input_hash_scalar;
+    unsigned char input_hash_local[32];
+    int overflow = 0;
+    int ret = 1;
+
+    /* Sanity check inputs */
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(public_data != NULL);
+    ARG_CHECK(outpoint_smallest36 != NULL);
+    ARG_CHECK((plain_pubkeys != NULL) || (xonly_pubkeys != NULL));
+    if (xonly_pubkeys != NULL) {
+        ARG_CHECK(n_xonly_pubkeys > 0);
+    } else {
+        ARG_CHECK(n_xonly_pubkeys == 0);
+    }
+    if (plain_pubkeys != NULL) {
+        ARG_CHECK(n_plain_pubkeys > 0);
+    } else {
+        ARG_CHECK(n_plain_pubkeys == 0);
+    }
+
+    /* Compute input public keys sum: A_sum = A_1 + A_2 + ... + A_n */
+    secp256k1_gej_set_infinity(&A_sum_gej);
+    for (i = 0; i < n_plain_pubkeys; i++) {
+        ret &= secp256k1_pubkey_load(ctx, &addend, plain_pubkeys[i]);
+        secp256k1_gej_add_ge_var(&A_sum_gej, &A_sum_gej, &addend, NULL);
```

Comment:
What happens if parsing fails and an `addend` other than expected (maybe invalid) is supplied to `secp256k1_gej_add_ge_var`?